### PR TITLE
Add support for building snaps

### PR DESCRIPTION
**Pull request by igorljubuncic**
Hi Filippo,

Mkcert looks like an interesting tool - I wish I had this back in 2008 when I was doing a lot of stuff around web servers. Oh well.

Anyway, I want to ask you to add support for snaps.

Now, snaps are cross-distro Linux software packages, supported on LTS and non-LTS Ubuntu, as well as many others distributions. Having mkcert as a snap would make it available to millions of users through the Snap Store (snapcraft.io/store), and also allow you to provide automatic updates with any changes to all your users. I guess this would make a lot of sense if there are vulnerabilities in the nss library, and people need updates fast, and you want to make sure everyone is fully aligned.

Snaps make for a really good use case here, as they are self-contained and isolated. In fact, I built mkcert as a strictly confined application, and if you decide to do snap building, you can play and tweak with whatthe snap can do some more.

TL;DR - technical details - for how to build snaps locally. I used Ubuntu 18.04 LTS as the build system.

snap install snapcraft --classic --beta
git clone https://github.com/igorljubuncic/mkcert.git
cd mkcert
git checkout add-snapcraft
snapcraft

The last command creates a <mkcert-version>.snap file, something like mkcert_1.0_amd64.snap.

This snap can be installed and tested locally with:

snap install mkcert_1.0_amd64.snap --dangerous

The --dangerous flag is necessary because the snap hasn’t yet gone through the snap store review process and is not digitally signed.

The mkcert command can be executed, e.g.: snap run mkcert <options>.

Then, you will need to get the snap actually published in the store:

- Register a developer account in the snap store https://snapcraft.io/account.
- Register the mkcert name in the store. 

snapcraft login
snapcraft register

- Upload a built snap to the store. The store supports multiple risk levels as “channels”. We have edge, beta, candidate and stable, which denote the level of stability. Typically, you start with edge and promote to other channels as you complete testing and validation.

snapcraft push mkcert_1.0_amd64.snap --release edge

- Test installing on a clean machine to see everything works as you expect.

snap install mkcert --edge

After you have completed your testing and you're happy, you can push a stable release to the stable channel. The next step after that is to update the store page, and promote the application online. We will gladly help there, and we'd be happy to feature your application in our store. Feel free to reach out if you have any questions or concerns.

### CODE REVIEW

File: snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  A simple zero-config tool to make locally trusted development certificates with any names you'd like.
+grade: stable
+confinement: strict
+base: core18
+parts:
+  mkcert:
+    plugin: go
+    source: https://github.com/FiloSottile/mkcert.git
+    go-importpath: github.com/FiloSottile/mkcert
+    build-packages:
```
Comment (adamdecaf) [inline snapcraft.yaml]: Is there an option to declare optional support for Java / openjdk also? Does it matter?

**Comment by igorljubuncic** [comment]
Hey Adam,

Here's a brief overview of how it works:

On the host system, you have the snapd service and snap userspace component - these are updated as part of your host's updates (say Ubuntu or Fedora).

Each snap is a self-contained squashfs filesystem containing snap (app) data, like bin, etc, lib, usr, var. Each snap comes with its own bundle and is not affected by libraries outside, so you could update your snap BEFORE the host system for example.

The snap store also has a mechanism for tracking libraries that you bundle with your snap, and when vulnerabilities are detected or reported for these, developers get an email informing them that their snaps contain out-of-date components, so they get update their snaps and rebuild them.

So for instance, you could have:

host system: /lib/example.1.0.so
snap1: /snap/foo/lib/example.1.1.so
snap2: /snap/bar/lib/example.1.7.so

And then if a vulnerability is detected in versions 1.0 and 1.1, the host and snap1 would need updates, but snap2 would continue working with the bundled library without being affected.

I hope this clarifies it.

File: snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+name: mkcert
+version: '1.0'
```
Comment (adamdecaf) [inline snapcraft.yaml]: This should match the latest released version, right? That's `1.1.2` as of today.

Comment (igorljubuncic) [reply snapcraft.yaml]: I used 1.0 as a first build tag, you can change it to whatever is most appropriate. In snapcraft, version does not carry a semantic meaning. So you can match it to your application version, git branch, etc.

**Comment by FiloSottile** [comment]
Hi, and thanks for opening a PR. While I understand what snaps are trying to do, I don't think they make for a sustainable dynamic in shifting the packaging effort to the upstream. mkcert is packaged for a dozen different package managers these days, and just documenting them all was a lot of maintenance overhead for a side project. Having to manage all package distributions would be unsustainable. If everyone were to actually standardize on snaps, it might make sense for us to maintain that one package, but in practice snaps are just one of a myriad of distribution vectors.

**Pull request closed without merging**